source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+            # it can be added to the queue
+            if new_artifact is not None:
+                # Also store its hash in the file database
+                file_info = FileInfoDatabase(self._database)
+                file_info.add_file_info(artifact.location,
+                                        new_artifact.hash)
+                new_artifacts.append(new_artifact)
+
+        elif artifact.state is Analysed:
+
+            # Work out whether this artifact needs to be
+            # included in the build or not - tag it as having
+            # been seen so the system knows it is Analysed
+            required = False
+            for definition in artifact.defines:
+                if definition in discovery:
+                    new_discovery[definition] = "Seen"
+                    required = True
+
+            # Assuming it is needed, check its
+            # dependencies to know what needs doing
+            if required:
+                compiled = [False]*len(artifact.depends_on)
+                for idep, dependency in enumerate(artifact.depends_on):
+                    if dependency in discovery:
+                        # Are the dependencies compiled?
+                        if discovery[dependency] == "Compiled":
+                            compiled[idep] = True
+                    else:
+                        # If the dependency isn't in the list at all yet
+                        # then add an entry so the system knows we are
+                        # expecting it later (for the above check)
+                        new_discovery[dependency] = "HeardOf"
+
+                # If the dependencies are satisfied (or there weren't
+                # any) then the file can be compiled now
+                if len(compiled) == 0 or all(compiled):
+                    for definition in artifact.defines:
+                        task = self._taskmap[(artifact.filetype,
+                                              artifact.state)]
+                        new_artifacts.extend(task.run([artifact]))
+                        new_discovery[definition] = "Compiled"
+                else:
+                    # If the dependencies weren't all satisfied then
+                    # back on the queue for another pass later
+                    new_artifacts.append(artifact)
+            else:
+                # If it wasn't required it could be later, so
+                # put it back on the queue, unless the target
+                # has been compiled, in which case it wasn't
+                # needed at all!
+                if discovery[self._target] != "Compiled":
+                    new_artifacts.append(artifact)
+
+        elif artifact.state is Compiled:
+            # Begin populating the list for linking
+            new_objects.append(artifact)
+            # But do not return a new artifact - this object
+            # is "done" as far as the processing is concerned
```

Comment:
Why do we put the object back on the queue if it's done? Surely it should be dropped on the floor once it's compiled?

Reply:
I'm unsure what you mean - the objects don't go back onto the queue - `new_objects` is used to accumulate the shared `objects` array (at the end of the method) and the `artifact` is not put back on the queue.  The `objects` array is eventually used to provide the list of objects to the linking step (just below)